clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2119,21 +2123,75 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
     CXXMethodDecl *InitMethod = getMethodByName(RecordDecl, MethodName);
     assert(InitMethod && "The accessor/sampler must have the __init method");
 
+    // Get access mode of accessor.
+    const auto *AccessorSpecializationDecl =
+        cast<ClassTemplateSpecializationDecl>(RecordDecl);
+    const TemplateArgument AccessModeArg =
+        AccessorSpecializationDecl->getTemplateArgs().get(2);
+
     // Don't do -1 here because we count on this to be the first parameter added
     // (if any).
     size_t ParamIndex = Params.size();
     for (const ParmVarDecl *Param : InitMethod->parameters()) {
       QualType ParamTy = Param->getType();
       addParam(BS, ParamTy.getCanonicalType());
-      if (ParamTy.getTypePtr()->isPointerType())
+      if (ParamTy.getTypePtr()->isPointerType()) {
         handleAccessorPropertyList(Params.back(), RecordDecl, BS.getBeginLoc());
+
+        llvm::APSInt ReadEnumVal;
+        // Add implicit attribute to parameter decl when it is a read only
+        // SYCL accessor.
+        getAccessModeReadVal(AccessModeArg, ReadEnumVal);
+        if (AccessModeArg.getAsIntegral() == ReadEnumVal)
```

Comment:
in fact, unless there is a call to the `getAccessModeReadVal` that I'm missing, it probably should just return 'bool' for this condition.

Reply:
Done